### PR TITLE
last code review: merge develop into master

### PR DESCRIPTION
**Pull request by GrayTurtle**
gonna be a lot of merge conflicts

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,8 +16,8 @@ All instructions here are relative to a LINUX environment. There will be some
 differences if you are working on a non-LINUX system. We will not support any
 other development environment.
 
-This project was built using JAVA 7. It is recommended that you install
-JAVA&nbsp;7 when working with this project.
+This project was built using JAVA 8. It is recommended that you install the
+STANDARD version of JAVA&nbsp;8 when working with this project.
```
Comment (KevinWorkman) [inline README.md]: Might want to be more specific (Oracle Java) and maybe provide a link.

Comment (KevinWorkman) [reply README.md]: And actually I'd recommend you put any special instructions up top in a big **SPECIAL INSTRUCTIONS** box that's impossible for anybody to miss.

File: README.md [hunk 2]
```diff
@@ -38,7 +38,7 @@ JAVA&nbsp;7 when working with this project.
 
        ```
        $ sh run_server.sh <team_id> <team_secret> <port> <persistent-dir>
-       $ sh run_client.sh <host> <port>
+       $ sh run_simple_gui_client.sh <host> <port>
```
Comment (KevinWorkman) [inline README.md:41]: Did you fix the issue with hardcoding the port?

File: src/codeu/chat/DBColumns.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Stores all of the column names for each table
+ * in the database.
+ * 
+ * Classes should not be instantiated.
+ * 
+ * @author malikg
+ */
+public class DBColumns {
```
Comment (KevinWorkman) [inline src/codeu/chat/DBColumns.java:11]: As previously discussed, might consider turning these into interfaces.

File: src/codeu/chat/DerbyStore.java [hunk 4]
```diff
@@ -16,47 +16,76 @@
...
+import codeu.chat.common.DerbyDatabaseInteractions;
+import codeu.chat.DBColumns;
+
 import java.io.File;
+import java.io.IOException;
 import java.sql.Connection;
 
-public class DerbyStore {
+/**
+ *
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:27]: Probably want to explain what this class is used for.

File: src/codeu/chat/DerbyStore.java [hunk 5]
```diff
@@ -16,47 +16,76 @@
...
-	private static final Logger.Log LOG = Logger.newLog(ServerMain.class);
+	private final String getMessageById = "SELECT * FROM " + messageTableName + " WHERE messageid = ?";
+	private final String getMessageCountById = "SELECT messageCount FROM " + userMessageCountTableName + " WHERE userid = ?";
 	
 	
+	private static final Logger.Log LOG = Logger.newLog(ServerMain.class);
+	
+	/**
+	 * Checks to see if the tables exist
+	 * 		If they do not then we, create
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:73]: nit: this formatting is a little weird. I'd try to stick with standard column widths.

File: src/codeu/chat/DerbyStore.java [hunk 6]
```diff
@@ -16,47 +16,76 @@
...
+	 * 		them and log it.
+	 * 
+	 * 		If they do exist, then we skip 
+	 * 		the table creation process to
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:77]: nit: You don't skip the table creation process to avoid an exception, you skip table creation because they're already created. :p

File: src/codeu/chat/DerbyStore.java [hunk 7]
```diff
@@ -16,47 +16,76 @@
...
+	 */
 	public DerbyStore() {
 	
 		try {
 			// Load the class necessary
 			Class.forName(driver).newInstance();
 			
 			// Checks to see if the database directory exist
-			File database = new File("testchatapp");
+			File database = new File("../testchatapp");
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:88]: The location of the database should probably be a parameter, that way you use a different database for testing and for production.

File: src/codeu/chat/DerbyStore.java [hunk 8]
```diff
@@ -74,18 +103,22 @@ public DerbyStore() {
...
 			
 			// Create the chat user table
-			stmt.execute("CREATE TABLE " + userTableName + "(id varchar(255), name varchar(255), password varchar(255), creation BIGINT)");
+			stmt.execute("CREATE TABLE " + userTableName + " (userid varchar(255), name varchar(255), password varchar(255), creation BIGINT, PRIMARY KEY (userid))");
 			
 			// Create the conversations table
-			stmt.execute("CREATE TABLE " + conversationTableName + "(id varchar(255), "
-					+ "owner varchar(255), creation BIGINT, title varchar(255), users varchar(255), firstMessage varchar(255)," +
-			"lastMessage varchar(255))");
+			stmt.execute("CREATE TABLE " + conversationTableName + " (conversationid varchar(255), "
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:109]: General comment: it would be cool to see a table diagram of your database in the design doc.

File: src/codeu/chat/DerbyStore.java [hunk 9]
```diff
@@ -95,33 +128,81 @@ public DerbyStore() {
 		}
 	}
 	
+	@Override
+	public User getUser(Uuid userid) throws SQLException, IOException {
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:132]: I'd add comments explaining what this returns, that it returns `null` if the user isn't found, and in what circumstances these exceptions will be thrown. Same for all of the other functions in this class.

File: src/codeu/chat/DerbyStore.java [hunk 10]
```diff
@@ -95,33 +128,81 @@ public DerbyStore() {
...
+		PreparedStatement getUser = conn.prepareStatement(getUserInfo);
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:133]: Still curious about the tradeoffs between creating a new `PreparedStatement` every time versus creating them all once at startup.

File: src/codeu/chat/DerbyStore.java [hunk 11]
```diff
@@ -95,33 +128,81 @@ public DerbyStore() {
...
+	@Override
+	public User userLogin(String name, String password) throws SQLException, IOException {
 		PreparedStatement checkValidUserTest = conn.prepareStatement(checkValidUser);
 		checkValidUserTest.setString(1, name);
 		checkValidUserTest.setString(2, password);
-		ResultSet user = checkValidUserTest.executeQuery();
+		ResultSet user = checkValidUserTest.executeQuery();	
 		
-		return (user != null) ? user.getString(1) : null;
+		return (user.next()) ? new User(Uuid.parse(user.getString(DBColumns.UserTable.userid)), user.getString(DBColumns.UserTable.name), user.getString(DBColumns.UserTable.password), Time.fromMs(user.getLong(DBColumns.UserTable.creation))) : null;
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:191]: This is a little complicated. I'd split it up into multiple lines, or add a comment explaining what it's doing.

File: src/codeu/chat/DerbyStore.java [hunk 12]
```diff
@@ -95,33 +128,81 @@ public DerbyStore() {
...
 	}
 	
+	@Override
 	public void addUser(User u) throws SQLException {
 		PreparedStatement addUser = conn.prepareStatement(addUserInfo);
 		addUser.setString(1, removeCharsInUuid(u.id.toString()));
 		addUser.setString(2, u.name);
-		addUser.setString(3, "test");
+		addUser.setString(3, u.password);
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:199]: Consider security: storing passwords as plaintext is a pretty big no-no. Might look into hashing if you're curious.

File: src/codeu/chat/DerbyStore.java [hunk 13]
```diff
@@ -224,23 +324,22 @@ public void updateLastMessage(Message m) throws SQLException {
 			
 			while (allConversationsResponse.next()) {
 				
-				Uuid conversationid = Uuid.fromString(removeCharsInUuid(allConversationsResponse.getString(1)));
+				Uuid conversationid = Uuid.parse(removeCharsInUuid(allConversationsResponse.getString("conversationid")));
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:327]: Can this use one of the column names in the constants classes?

File: src/codeu/chat/DerbyStore.java [hunk 14]
```diff
@@ -283,35 +383,56 @@ public void updateLastMessage(Message m) throws SQLException {
...
+		
+		if (messageCount.next()) {
+			return messageCount.getInt(DBColumns.UserMessageCountTable.messageCount);
+		}
+		
+	return 0;
+	}
+
+	@Override
+	public int setUserMessageCount(Uuid userid) throws SQLException {
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:401]: I'd maybe rename this to `incrementUserMessageCount()` or something. And add comments explaining exactly what it does.

File: src/codeu/chat/DerbyStore.java [hunk 15]
```diff
@@ -283,35 +383,56 @@ public void updateLastMessage(Message m) throws SQLException {
...
-	// them, and decreases the different numbers in 
-	// the uuid since it could be larger than an int.
+	// and closing brackets.
 	public String removeCharsInUuid(String uuid) {
 		// Remove the characters when uuid is transformed to a string
 		uuid = uuid.replace("[", "");
 		uuid = uuid.replace("]", "");
 		uuid = uuid.replace("UUID:", "");
 		
-		// Prevent a number larger than the max possible int
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:297]: Did one of us delete these lines or is this coming from the admins?

File: src/codeu/chat/client/ClientConversation.java [hunk 16]
```diff
@@ -44,6 +48,41 @@
...
+   * mapped to (newest to oldest) when inserted into the TreeMap
+   * summariesSortedByCreationTime.
+   * @author malikg
+   *
+   */
+  class sortByCreation implements Comparator<ConversationSummary> {
+	  @Override
+	  public int compare(ConversationSummary a, ConversationSummary b) {
+		  
+		  if (!Uuid.equals(a.lastMessage, Uuid.NULL) && !Uuid.equals(b.lastMessage, Uuid.NULL)) {
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientConversation.java:68]: I'd add a comment explaining what this comparison logic is doing.

File: src/codeu/chat/client/ClientConversation.java [hunk 17]
```diff
@@ -164,14 +203,23 @@ private void updateCurrentConversation() {
   }
 
   public int conversationsCount() {
-   System.out.println(this.getClass().toString() + " conversationsCount()");
+	  System.out.println(this.getClass().toString() + " conversationsCount()");
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientConversation.java:206]: You probably don't want `System.out.println()` statements in production code. Either use the log or delete them.

File: src/codeu/chat/client/ClientConversation.java [hunk 18]
```diff
@@ -180,10 +228,13 @@ public void updateAllConversations(boolean currentChanged) {
 	  System.out.println(this.getClass().toString() + " updateAllConversations()");
     summariesByUuid.clear();
     summariesSortedByTitle = new Store<>(String.CASE_INSENSITIVE_ORDER);
+    
+    summariesSortedByCreationTime = new TreeMap<ConversationSummary, String>(new sortByCreation());
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientConversation.java:232]: What if two `ConversationSummary` instances happen to be created at the same exact time?

File: src/codeu/chat/client/ClientUser.java [hunk 19]
```diff
@@ -44,17 +45,17 @@ public ClientUser(Controller controller, View view) {
...
+  /**
+   * Takes in a username or password entry and checks to make sure it's only made
+   * up of alphanumeric characters.
+   * @param userInput
+   * @return
+   */
+  public static boolean isValidInput(String userInput) {
 
-      // TODO: check for invalid characters
+    boolean validInput = Pattern.matches("[a-zA-Z0-9]+", userInput);
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientUser.java]: I'd add a comment explaining exactly what this regex means.

File: src/codeu/chat/client/ClientUser.java [hunk 20]
```diff
@@ -150,4 +166,12 @@ public String showUserInfo(String uname) {
   public static void printUser(User user) {
     System.out.println(getUserInfoString(user));
   }
+
+  public int getMessageCount(Uuid userid) {
+      return view.getMessageCount(userid);
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientUser.java:182]: It seems a little weird that one of the `getMessageCount()` function goes to the view, but the `increaseMessageCount()` function goes to the controller. I'd expect things to be one-way: the view calls the controller, and the controller calls the model. Seems a little weird to have something going to both the view and the controller.

File: src/codeu/chat/client/Controller.java [hunk 21]
```diff
@@ -112,4 +116,27 @@ public Conversation newConversation(String title, Uuid owner)  {
 
     return response;
   }
+  
+  public int updateMessageCount(Uuid userid) {
```
Comment (KevinWorkman) [inline src/codeu/chat/client/Controller.java:120]: I'd add comments explaining what this is updating. This is taking an already updated count and sending it to the server, right?

File: src/codeu/chat/client/View.java [hunk 22]
```diff
@@ -280,4 +280,71 @@ public Uuid getUserGeneration() {
 
     return messages;
   }
+
+  public User checkUserLogin(String name, String password) {
```
Comment (KevinWorkman) [inline src/codeu/chat/client/View.java:284]: Shouldn't these types of functions be in the controller?

File: src/codeu/chat/client/commandline/Chat.java [hunk 23]
```diff
@@ -89,7 +89,7 @@ private void doOneCommand(Scanner lineScanner) {
       if (!tokenScanner.hasNext()) {
         System.out.println("ERROR: No user name supplied.");
       } else {
-        signInUser(tokenScanner.nextLine().trim());
+        signInUser(tokenScanner.nextLine().trim(), ""); // Note: putting an empty string for password because we're not using command line
```
Comment (KevinWorkman) [inline src/codeu/chat/client/commandline/Chat.java:92]: If you're not using the command line, you should probably just delete these classes.

File: src/codeu/chat/client/simplegui/ChatGuiFX.java [hunk 24]
```diff
@@ -1,120 +1,231 @@
...
-            LOG.error(ex, "Exception in ChatGuiFX.run");
-            System.exit(1);
+    private static final double WINDOW_WIDTH = 1000;
+    private static final double WINDOW_HEIGHT = 500;
+    private static final String SIGNIN_ERROR_MESSAGE = "Your username or password does not match. Have you signed up yet?";
+    private static final String SIGNUP_ERROR_MESSAGE = "Sorry, that username already exists. Please choose a different one.";
+    private static final String BADCHAR_ERROR_MESSAGE = "Usernames and passwords can only be composed of letters and numbers, no special characters.";
+
+    // both page vars
+    //private ClientContext clientContext;
```
Comment (KevinWorkman) [inline src/codeu/chat/client/simplegui/ChatGuiFX.java]: Can probably get rid of commented-out code.

File: src/codeu/chat/client/simplegui/ChatGuiFX.java [hunk 25]
```diff
@@ -1,120 +1,231 @@
...
 
-        signInPane.setTop(signInLabelPane);
-        signInPane.setCenter(signInButton);                 // Add buttons and labels to panes
+    public void setContext(Controller controller, View view) {
+    	clientContext = new ClientContext(controller, view);
+    }
+
+    public void launch(Controller controller, View view) {
+    	setContext(controller, view);
+        Application.launch(ChatGuiFX.class);
```
Comment (KevinWorkman) [inline src/codeu/chat/client/simplegui/ChatGuiFX.java]: I'm still suspicious of this line. Isn't this creating a new instance of `ChatGuiFX`, when we're already inside an instance? Doesn't this mean you end up with two instances?

File: src/codeu/chat/client/simplegui/ChatGuiFX.java [hunk 26]
```diff
@@ -1,120 +1,231 @@
...
+    }
 
+    @Override
+    public void start(Stage primaryStage) throws Exception {
```
Comment (KevinWorkman) [inline src/codeu/chat/client/simplegui/ChatGuiFX.java]: Why does this throw an Exception?

File: src/codeu/chat/client/simplegui/ChatGuiFX.java [hunk 27]
```diff
@@ -1,120 +1,231 @@
...
+        inputMasterBox.getChildren().add(buttonBox);
 
+        signInPane.setTop(signInLabelPane);
+        signInPane.setBottom(errorLabel);
+        // Add labels and input box to the pane
+        signInPane.setCenter(inputMasterBox);
 
+        signInScene = new Scene(signInPane, WINDOW_WIDTH, WINDOW_HEIGHT);
 
         // Main Page
```
Comment (KevinWorkman) [inline src/codeu/chat/client/simplegui/ChatGuiFX.java]: If at all possible, you should probably split this into separate classes, or at least separate functions.

File: src/codeu/chat/client/simplegui/ChatGuiFX.java [hunk 28]
```diff
@@ -138,21 +249,289 @@ public void start(Stage primaryStage) throws Exception {
...
+            }
+        }
+    }
+
+    // TODO: set up a loop to where this updates every half sec or so
+    /**
+    * When the user presses the update button, update the GUI by clearing & and filling
+    * all the lists that will be displayed with updated content
+    */
+    private void updateGUI(ActionEvent e) {
```
Comment (KevinWorkman) [inline src/codeu/chat/client/simplegui/ChatGuiFX.java]: Since you're clearing the GUI out and completely rebuilding it (possible every half second), think about what would happen if you had a million (or more!) users, conversations, messages, etc. Is there a way to avoid rebuilding everything every time?

File: src/codeu/chat/client/simplegui/ChatGuiFX.java [hunk 29]
```diff
@@ -138,21 +249,289 @@ public void start(Stage primaryStage) throws Exception {
...
+        clientContext.message.updateMessages(true);
+        fillMessagesList(clientContext.conversation.getCurrent());
+    }
+
+    // TODO: add more colors
+    /**
+    * Colorizes user based on how many messages has been sent
+    * @param userID  the ID of the user to have their name colorized
+    */
+    private void colorizeUsername(Uuid userID) {
```
Comment (KevinWorkman) [inline src/codeu/chat/client/simplegui/ChatGuiFX.java]: It's a little unclear exactly what this function is for. Where is `userName` coming from? Does this style every user, or just the current user? Can I see other users' colors?

File: src/codeu/chat/client/simplegui/UserPanel.java [hunk 30]
```diff
@@ -151,7 +151,7 @@ public void actionPerformed(ActionEvent e) {
       public void actionPerformed(ActionEvent e) {
         if (userList.getSelectedIndex() != -1) {
           final String data = userList.getSelectedValue();
-          clientContext.user.signInUser(data);
+          clientContext.user.signInUser(data, "");  // Putting in empty string because we don't use the swing gui
```
Comment (KevinWorkman) [inline src/codeu/chat/client/simplegui/UserPanel.java:154]: You can probably just delete these classes then.

File: src/codeu/chat/common/DerbyDatabaseInteractions.java [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+	public Store<Uuid, User> getAllUsers();
+	
+	/**
+	 * 
+	 * @param userid
+	 * @return a specific user stored stored in the database.
+	 * @throws SQLException
+	 * @throws IOException
+	 */
+	public User getUser(Uuid userid) throws SQLException, IOException;
```
Comment (KevinWorkman) [inline src/codeu/chat/common/DerbyDatabaseInteractions.java:30]: I'm still curious about why the interface is declaring thrown exceptions. Do these exceptions ever actually happen? Could they be isolated in the implementing class? This is requiring you to use a bunch of try-catch blocks that don't actually do anything when an exception happens. What if I want to create an implementation of this interface that uses XML files instead of sql?

File: src/codeu/chat/server/Model.java [hunk 32]
```diff
@@ -96,6 +98,52 @@ public void add(User user) {
     userByTime.insert(user.creation, user);
     userByText.insert(user.name, user);
   }
+  
+  public User checkUserLogin(String name, String password) {
```
Comment (KevinWorkman) [inline src/codeu/chat/server/Model.java:102]: I'd add comments to these functions.

File: src/codeu/chat/server/Model.java [hunk 33]
```diff
@@ -96,6 +98,52 @@ public void add(User user) {
...
+	  User user = null;
+	  try {
+		  user = ds.userLogin(name, password);
+	  }
+	  catch (Exception ex) {
```
Comment (KevinWorkman) [inline src/codeu/chat/server/Model.java:107]: When does this exception actually happen? Should you communicate something back to the user? Same comment for other exceptions in this class.

File: src/codeu/chat/server/Server.java [hunk 34]
```diff
@@ -241,8 +245,67 @@ private boolean onMessage(InputStream in, OutputStream out) throws IOException {
...
+    } else if (type == NetworkCode.CHECK_USER_REQUEST) {
+    	
+    	final String username = Serializers.STRING.read(in);
+    	final String password = Serializers.STRING.read(in);
+    	
+    	User validUser = null;
+    	try {
+    		validUser = model.checkUserLogin(username, password);
+    	}
+    	catch (Exception ex) {
```
Comment (KevinWorkman) [inline src/codeu/chat/server/Server.java:257]: When will this exception happen? Same comment for the other exceptions in this class.

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 35]
```diff
@@ -0,0 +1,230 @@
...
+import codeu.chat.common.Message;
+
+/**
+ * 
+ * @author malikg
+ *
+ */
+public class DerbyStoreTest {
+
+	private DerbyStore ds;
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java:23]: I would sorta expect these to each have their own tests classes: one that tests the low-level sql, another that tests the model functions, and another that tests the controller functions. Each of them would only use one class at a time and use hardcoded input and output for everything else.

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 36]
```diff
@@ -0,0 +1,230 @@
...
+
+		assertTrue(
+				"Check that messages can be retrieved and have the same id",
+				Uuid.equals(message.next, nextMessage.id));
+
+	}
+	
+	@Test
+	public void checkUserLogin() {
+		String username = "bob" + Time.now();
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java:190]: I'm guessing you had to do this because you're using the same database for testing every time. Instead, you probably want to clear out the database when you're done testing. You'd also want to use a different database for testing and for production.

File: README.md (hunk 1 above)
Comment (KevinWorkman) [reply README.md]: Do you still have to use port `2007`? Make sure that's in here too.

And while we're at it, you might want to throw in some screenshots, especially of your new features. Also include a big "hey! try sending X number of messages to see your username turn red!" just so people know exactly what you've done.

**Review by KevinWorkman — COMMENTED** [review]
Awesome to see tests and comments coming in. There are four main things you're going to hear about:

- **Documentation:** there are still a bunch of uncommented functions.

- **Separation of model, view, and controller:** I'd expect the view to call the controller, and the controller to call the model (database). Right now there is quite a bit of blurring between all the different layers.

- **Tests:** Similar to above, I'd usually expect a test class for the controller, and a separate test class for the model. I'd also like to see more tests for corner cases, not just normal use cases.

- **Scalability:** This is Google's favorite question. What would happen if you had a million users, with a million conversations, which had a million messages in each? How would your database have to change? (What if it was too large to fit on a single hard drive?) How would your data structures have to change? (What if it no longer fit into memory?) How would your algorithms have to change? (For loops are evil.)

Other than that, I'd just make sure that things are obvious: somebody who has never looked at your code before needs to know how to run it and how to try out your new features. Make this as obvious as possible, in a big impossible-to-miss box at the top of your readme, as well as in your design doc.
